lxd/instance/drivers/driver_qemu.go

```diff
@@ -4686,6 +4689,43 @@ func (d *qemu) Stop(stateful bool) error {
 		return fmt.Errorf("Stateful stop requires migration.stateful to be set to true")
 	}
 
+	pool, err := d.getStoragePool()
+	if err != nil {
+		return err
+	}
+
+	sharedStorage := pool.Driver().Info().Remote
+	if sharedStorage {
+		_, rootDiskDevice, err := d.getRootDiskDevice()
+		if err != nil {
+			return err
+		}
+
+		stateDiskSizeStr := pool.Driver().Info().DefaultVMBlockFilesystemSize
+		if rootDiskDevice["size.state"] != "" {
+			stateDiskSizeStr = rootDiskDevice["size.state"]
+		}
+
+		stateDiskSize, err := units.ParseByteSizeString(stateDiskSizeStr)
+		if err != nil {
+			return err
+		}
+
+		memoryLimitStr := QEMUDefaultMemSize
+		if d.expandedConfig["limits.memory"] != "" {
+			memoryLimitStr = d.expandedConfig["limits.memory"]
+		}
+
+		memoryLimit, err := units.ParseByteSizeString(memoryLimitStr)
+		if err != nil {
+			return err
+		}
+
+		if stateDiskSize < memoryLimit {
+			return fmt.Errorf("Stateful stop requires that the instance limits.memory is less than size.state on the root disk device")
```

Comment:
Lets make this error more generic so it be be used in both start and stop scenarios, something like:

> When migration.stateful is enabled the root disk's size.state setting should be set to at least the limits.memory size in order to accommodate the stateful stop file